code/game/machinery/dance_machine.dm

```diff
@@ -153,60 +97,90 @@
-					if(!COOLDOWN_FINISHED(src, jukebox_error_cd))
-						return
-					playsound(src, 'sound/misc/compiler-failure.ogg', 50, TRUE)
-					COOLDOWN_START(src, jukebox_error_cd, 15 SECONDS)
-					return
+			if(isnull(music_player.active_song_sound))
+				if(!COOLDOWN_FINISHED(src, jukebox_song_cd))
+					to_chat(usr, span_warning("Error: The device is still resetting from the last activation, \
+						it will be ready again in [DisplayTimeText(COOLDOWN_TIMELEFT(src, jukebox_song_cd))]."))
+					if(COOLDOWN_FINISHED(src, jukebox_error_cd))
+						playsound(src, 'sound/misc/compiler-failure.ogg', 33, TRUE)
+						COOLDOWN_START(src, jukebox_error_cd, 15 SECONDS)
+					return TRUE
+
 				activate_music()
-				START_PROCESSING(SSobj, src)
-				return TRUE
 			else
-				stop = 0
-				return TRUE
+				stop_music()
+
+			return TRUE
+
 		if("select_track")
-			if(active)
+			if(!isnull(music_player.active_song_sound))
 				to_chat(usr, span_warning("Error: You cannot change the song until the current one is over."))
-				return
-			var/list/available = list()
-			for(var/datum/track/S in songs)
-				available[S.song_name] = S
-			var/selected = params["track"]
-			if(QDELETED(src) || !selected || !istype(available[selected], /datum/track))
-				return
-			selection = available[selected]
+				return TRUE
+
+			var/datum/track/new_song = music_player.songs[params["track"]]
+			if(QDELETED(src) || !istype(new_song, /datum/track))
+				return TRUE
+
+			music_player.selection = new_song
 			return TRUE
+
 		if("set_volume")
 			var/new_volume = params["volume"]
-			if(new_volume == "reset")
-				volume = initial(volume)
-				return TRUE
+			if(new_volume == "reset" || new_volume == "max")
+				music_player.set_volume_to_max()
 			else if(new_volume == "min")
-				volume = 0
-				return TRUE
-			else if(new_volume == "max")
-				volume = initial(volume)
-				return TRUE
-			else if(text2num(new_volume) != null)
-				volume = text2num(new_volume)
-				return TRUE
+				music_player.set_new_volume(0)
+			else if(isnum(text2num(new_volume)))
+				music_player.set_new_volume(text2num(new_volume))
+			return TRUE
+
+		if("loop")
+			music_player.sound_loops = !!params["looping"]
+			return TRUE
 
 /obj/machinery/jukebox/proc/activate_music()
-	active = TRUE
+	if(!isnull(music_player.active_song_sound))
+		return FALSE
+
+	music_player.start_music()
 	update_use_power(ACTIVE_POWER_USE)
 	update_appearance(UPDATE_ICON_STATE)
-	START_PROCESSING(SSobj, src)
-	stop = world.time + selection.song_length
+	if(!music_player.sound_loops)
+		song_timerid = addtimer(CALLBACK(src, PROC_REF(stop_music)), music_player.selection.song_length, TIMER_UNIQUE|TIMER_STOPPABLE|TIMER_DELETE_ME)
+	return TRUE
 
 /obj/machinery/jukebox/disco/activate_music()
-	..()
+	. = ..()
+	if(!.)
+		return
 	dance_setup()
 	lights_spin()
+	begin_processing()
+
+/obj/machinery/jukebox/proc/stop_music()
```

Comment:
why is this in the middle of the code for the `disco` subtype? should be moved to be with the rest of the base-type code.